src/main/java/at/ac/tuwien/kr/alpha/common/Interner.java

```diff
@@ -0,0 +1,26 @@
+package at.ac.tuwien.kr.alpha.common;
+
+import java.lang.ref.WeakReference;
+import java.util.WeakHashMap;
+
+public class Interner<T> {
+	private WeakHashMap<T, WeakReference<T>> pool = new WeakHashMap<>();
+
+	public synchronized T intern(T object) {
+		T res;
```

Comment:
* [Java Object Interning](https://www.javacodegeeks.com/2014/03/java-object-interning.html) (this is where the code comes from)
 * [Understanding Weak References](https://community.oracle.com/blogs/enicholas/2006/05/04/understanding-weak-references)

Interners allow for reference equality checks, and (maybe) save some memory.

In essence, we are using weak references here, because there is no point in keeping objects interned when there are no other references to it. If the objects we are referring to are not (strongly) referred by anything else, then they might just "disappear" from the interner.

Conversely, if we would not use weak references, then we would need additional logic that decides when an interned object should be removed from the interner, e.g. we would have to track when for example an instance of `ConstantTerm` can be garbage collected. We do not want that.